### PR TITLE
Purchases: Handle asynchronous plan removals/cancellations from Atomic sites

### LINKED ISSUE #55194: Atomic Downgrade: Support queued cancellation/removals
### Details

D66763-code adds support for asynchronously removing/cancelling a plan from an Atomic site, so let's remove the current restriction in Calypso that prevents users for performing that action.

### Requirements

- Place logic behind an inactive [feature flag](https://github.com/Automattic/wp-calypso/tree/trunk/config#config). Should be explicitly enabled using the `?flags=` param.
- Render current Simple site plan pre-cancellation form after clicking on the remove/cancel button. Do not display any additional plugins/themes loss warning at the moment, this will be done in a follow-up.
- The remove/cancel button should be disabled if the site is reverting. In a follow-up we can improve this by showing extra information on a dialog.

### PR DESCRIPTION
**Pull request by mmtr**
Fixes https://github.com/Automattic/wp-calypso/issues/55194

#### Changes proposed in this Pull Request

Screen Shot 2021-09-15 at 14 00 11

Implements a basic proof of concept that handles the asynchronous removals/cancellations of plans from Atomic sites.

The PoC lives behind the `atomic/automated-revert` feature flag which is disabled by default in all environments.

#### Testing instructions

- Use the Calypso live link below.
- Switch to an Atomic site.
- Go to Upgrades > Purchases.
- Select the site plan.
- Try removing/cancelling it.
- Make sure you are asked to contact support in order to remove/cancel it.
- Append `?flags=atomic/automated-revert` at the end of the URL and reload.
- Go through the removal/cancellation flow.
- Make sure you get a success notification indicating the removal/cancellation has been queued.
- Make sure you cannot remove/cancel the plan again (the button shouldn't show up).
- Manually run the `atomic_subscription_revert_after_lossless_import` async job (see testing instructions in D66763-code).
- Reload Calypso.
- Make sure the Atomic site is on a Free plan now.
- Make sure there are no regression when removing/cancelling a plan from a Simple site (the success notification should denote the plan has been immediately removed/cancelled).

### CODE REVIEW

**Comment by BogdanUngureanu** [comment]
> Make sure you cannot remove/cancel the plan again (the button shouldn't show up).

What if we make it disabled instead and display a tooltip explaining to the user that their plan is in the process of cancellation? 🤔

**Comment by mmtr** [comment]
> What if we make it disabled instead and display a tooltip explaining to the user that their plan is in the process of cancellation? 🤔

I considered that, but then remembered that tooltips are not accessibility friendly (although doable with some additional work: paYJgx-WR-p2), so I decided to take the simplest approach here.

If we receive reports of users being confused, we can create a follow-up issue and ask the design team to explore some alternatives.

File: client/state/purchases/selectors/should-revert-atomic-site-before-deactivation.js [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+	if ( ! isSiteAutomatedTransfer( state, purchase.siteId ) ) {
+		return false;
+	}
+
+	// Retrieves all Atomic supported purchases for the site.
+	// If there is at least one active subscription except the given one,
+	// the site needs to be kept in the Atomic infra.
+	return getSitePurchases( state, purchase.siteId ).some(
+		( sitePurchase ) =>
+			sitePurchase.siteId !== purchaseId && isAtomicSupportedProduct( purchase.productSlug )
```
Comment (BogdanUngureanu) [inline client/state/purchases/selectors/should-revert-atomic-site-before-deactivation.js]:
```
sitePurchase.**siteId** !== purchaseId
```
Is this intended? :)

**Comment by BogdanUngureanu** [comment]
Great work, Miguel! I've followed the steps and the async process worked great! However, I've noticed that navigating though the interface, the cancel button is removed on all subscriptions (non-Atomic) and on a different site:

https://user-images.githubusercontent.com/995955/133623011-a4202f9f-4c11-4793-9f19-a97f5f16f31c.mov

File: client/state/purchases/selectors/should-revert-atomic-site-before-deactivation.js (hunk 1 above)
Comment (mmtr) [reply client/state/purchases/selectors/should-revert-atomic-site-before-deactivation.js]: Good catch! Fixed in https://github.com/Automattic/wp-calypso/pull/55381/commits/7d2d9d70abbd1c729099b7d2c666f34f7473b227

**Comment by mmtr** [comment]
> navigating though the interface, the cancel button is removed on all subscriptions (non-Atomic) and on a different site

Oops! I think I covered that in https://github.com/Automattic/wp-calypso/commit/b79e7098327b5279c93d060aa1fc76616106d708, can you confirm?

**Comment by mmtr** [comment]
> Oops! I think I covered that in [b79e709](https://github.com/Automattic/wp-calypso/commit/b79e7098327b5279c93d060aa1fc76616106d708), can you confirm?
Actually, it doesn't completely fix the issue. I'll take a further look.

**Comment by dsas** [comment]
Removing simple atomic sites still works ✅
After cancelling a site, I get a success notification but the button still appears ❎:
https://user-images.githubusercontent.com/93301/134016496-cd81d370-6a7e-4521-b2d2-b3241aee0b68.mp4

The cancel API returned 
```
{"body":{"status":"queued","message":"Your purchase has been scheduled for cancellation."},"status":200,"headers":{"Allow":"POST, PUT, PATCH"}}
```

The stickers API response included `subscription_deactivation_locked`

Edit: I should point out that everything else worked after that too.

**Comment by mmtr** [comment]
> After cancelling a site, I get a success notification but the button still appears ❎:

Really good catch! I had quite a bug in the logic that determines whether the site should be reverted which was also affecting other parts of the flow 😅. Should've been fixed in https://github.com/Automattic/wp-calypso/pull/55381/commits/50680390cba0d0b5dbcc3dbd8282ea2a32021f89

**Comment by BogdanUngureanu** [comment]
Hey, I've just tested it again and noticed that for the Jetpack Search add-on on a free AT plan the user is unable to self cancel the subscription and required to contact us:
Screenshot 2021-09-21 at 12 34 29

Also, about the /me/purchases issue, I've noticed that the bug is reproduced on Production as well. I'll open a separate issue for it because I believe it's unrelated to your implementation.

**Comment by mmtr** [comment]
> that for the Jetpack Search add-on on a free AT plan the user is unable to self cancel the subscription and required to contact us

Nice finding! Solved in https://github.com/Automattic/wp-calypso/pull/55381/commits/6b74b3825911454ce2253ba0b7f4c637efd2cb8a

> about the /me/purchases issue

I'm sorry, what's that issue? I think I'm missing something.

**Comment by dsas** [comment]
Simple site cancels still work ok ✅ 
The button to remove & refund is now removed once used ✅ 

I'm not sure how much of a problem this will be in practice as it is an unlikely scenario and things will hopefully move faster most of the time but it seems that there is an issue if you remove & refund the subscription, alter the subscription and then run the queued job.

Here is the step-by-step:
1. Get an atomic site
2. Remove & refund, get 'Your purchase has been scheduled for cancellation' message 
3. Look at the purchase again, notice the remove & refund buttons correctly no longer appear, also notice it says 'auto-renew is on and I will be billed on date', which might confuse/worry people.
4. Click Renew New and renew
5. Manually create and run `atomic_subscription_revert_after_lossless_import`
6. Can't remove & refund the new purchase
7. Refresh page
8. Purchase has disappeared, site is free.

In store-admin, I can see under both payments and blog transactions that I bought a business plan and then bought a renewal. I can see that the renewal was refunded but not the initial payment.

This is probably more of a problem for D66763-code but if the front-end blocked access to renewals etc, that would reduce the likelihood of things happening, and reduce confusion for the customer.

**Comment by BogdanUngureanu** [comment]
> I'm sorry, what's that issue? I think I'm missing something.

The one from this video: https://github.com/Automattic/wp-calypso/pull/55381#issuecomment-920915865 To be more precise:
- Go to /me/purchases
- The items (sites) in the list jump around (another minor issue)
- Click on your Atomic site with an in progress subscription cancellation
- Go back to /me/purchases (without refreshing the page)
- Click on a different Atomic site
- Notice that the cancel button is not visible
- Refresh the page and you'll be redirected from subscription page to the main /me/purchases page.

From what I've noticed, the ID's in the /me/purchases somehow pass a different subscription id (it's possible it's a side effect of the store sandbox).

@dsas I'll look into it. I think we should also disable the `Upgrade` button for Business plans when the async cancellation flow is executed.

**Comment by mmtr** [comment]
> if the front-end blocked access to renewals etc, that would reduce the likelihood of things happening, and reduce confusion for the customer.

Agreed! Done in https://github.com/Automattic/wp-calypso/pull/55381/commits/bb92694c2805e00428f84468612164a74b0db5a0.

> the ID's in the /me/purchases somehow pass a different subscription id 

I think I fixed that in https://github.com/Automattic/wp-calypso/pull/55381/commits/bb92694c2805e00428f84468612164a74b0db5a0#diff-fe5caece9a3dd7aa5c8903939e0224272361780ce2d743016bdaf46c81c086caR836

**Comment by BogdanUngureanu** [comment]
> I think I fixed that in bb92694#diff-fe5caece9a3dd7aa5c8903939e0224272361780ce2d743016bdaf46c81c086caR836

Looks like it was from the sandbox, I wasn't able to reproduce it un-sandboxed.

Works great, I wasn't able to find any regression. 🎉 

There was an edge case, but I think it's non-blocking: If you remove the feature flag on a site that is pending for subscription deletion, the buttons (e.g. upgrade, renew, etc) are still removed.

Let's see if Dean notices anything else 🤞 .

**Review by BogdanUngureanu — APPROVED** [review]

**Comment by dsas** [comment]
Excellent 👍 this is working well. It does say 'Auto-renew is ON' which is technically correct but perhaps not helpful and for me, displays a notification about my expiring card. I think this is acceptable though, particularly when e.g. #56421 is implemented.

**Review by dsas — APPROVED** [review]

**Comment by a8ci18n** [comment]
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/6640640

Thank you @mmtr for including a screenshot in the description! This is really helpful for our translators.

**Comment by a8ci18n** [comment]
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/6640640).

**Pull request merged**